File: packages/sdk-python/src/swarmdock/crypto.py
"""Ed25519 key generation and signing for SwarmDock authentication."""

import base64
from nacl.signing import SigningKey, VerifyKey


def generate_keypair() -> dict[str, str]:
    """Generate an Ed25519 keypair. Returns base64-encoded private and public keys."""
    signing_key = SigningKey.generate()
    return {
        "private_key": base64.b64encode(signing_key.encode() + signing_key.verify_key.encode()).decode(),
        "public_key": base64.b64encode(signing_key.verify_key.encode()).decode(),
    }


def sign_message(private_key_b64: str, message: str) -> str:
    """Sign a message with an Ed25519 private key. Returns base64-encoded signature."""
    key_bytes = base64.b64decode(private_key_b64)
    signing_key = SigningKey(key_bytes[:32])
    signed = signing_key.sign(message.encode())
    return base64.b64encode(signed.signature).decode()


def get_public_key(private_key_b64: str) -> str:
    """Extract the public key from a private key. Returns base64-encoded public key."""
    key_bytes = base64.b64decode(private_key_b64)
    signing_key = SigningKey(key_bytes[:32])
    return base64.b64encode(signing_key.verify_key.encode()).decode()

File: packages/sdk-go/crypto.go
package swarmdock

import (
	"crypto/ed25519"
	"crypto/rand"
	"encoding/base64"
)

// GenerateKeypair creates a new Ed25519 keypair. Returns base64-encoded private and public keys.
func GenerateKeypair() (privateKeyB64, publicKeyB64 string) {
	pub, priv, _ := ed25519.GenerateKey(rand.Reader)
	return base64.StdEncoding.EncodeToString(priv), base64.StdEncoding.EncodeToString(pub)
}

// SignMessage signs a message with an Ed25519 private key. Returns base64-encoded signature.
func SignMessage(privateKeyB64, message string) (string, error) {
	keyBytes, err := base64.StdEncoding.DecodeString(privateKeyB64)
	if err != nil {
		return "", err
	}
	priv := ed25519.PrivateKey(keyBytes)
	sig := ed25519.Sign(priv, []byte(message))
	return base64.StdEncoding.EncodeToString(sig), nil
}

// GetPublicKey extracts the public key from a private key. Returns base64-encoded public key.
func GetPublicKey(privateKeyB64 string) (string, error) {
	keyBytes, err := base64.StdEncoding.DecodeString(privateKeyB64)
	if err != nil {
		return "", err
	}
	priv := ed25519.PrivateKey(keyBytes)
	pub := priv.Public().(ed25519.PublicKey)
	return base64.StdEncoding.EncodeToString(pub), nil
}
